src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,177 @@
+    return tokenPrices[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache = this.getPriceCache(currency);
+
+    // Determine whether *all* prices are current.
+    const now = msToS(Date.now());
+    const missed: { [address: string]: number } = {};
+    addresses.forEach((address: string) => {
+      const addr = address.toLowerCase();
+      // note: If priceCache[addr] is undefined, it is assigned here.
+      const tokenPrice = (priceCache[addr] = priceCache[addr] ?? ({ price: 0, timestamp: 0 } as TokenPrice));
+      const age = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_SAFE_INTEGER;
+
+      if (age > this.maxPriceAge) {
+        missed[address] = age;
+      }
+    });
+
+    if (Object.keys(missed).length > 0) {
+      const requestAddresses = Object.keys(priceCache);
+      this.logger.debug({
+        at: "PriceClient#getPricesByAddress",
+        message: `${currency.toUpperCase()} cache miss (age > ${this.maxPriceAge} S).`,
+        tokens: missed,
+      });
+      const prices = await this.requestPrices(requestAddresses, currency);
+      if (Object.keys(prices).length === 0) {
+        this.logger.warn({ at: "PriceClient#getPricesByAddress", message: "Failed to update token prices." });
+        // @todo throw ?
+        return [];
+      }
+      this.updateCache(priceCache, prices, requestAddresses);
+    }
+
+    return addresses.map((addr: string) => {
+      const { price, timestamp } = priceCache[addr.toLowerCase()];
+      return { address: addr, price, timestamp };
+    });
+  }
+
+  expireCache(currency: string): void {
+    const priceCache = this.getPriceCache(currency);
+    Object.values(priceCache).forEach((token: TokenPrice) => (token.timestamp = 0));
+    this.logger.debug({ at: "PriceClient#expireCache", message: `Expired ${currency} cache.` });
+  }
+
+  protected getPriceCache(currency: string): PriceCache {
+    if (this.prices[currency] === undefined) this.prices[currency] = {};
+    return this.prices[currency];
+  }
+
+  private async requestPrices(addresses: string[], currency: string): Promise<PriceCache> {
+    let prices: TokenPrice[] = [];
+
+    for (const priceFeed of this.priceFeeds) {
+      this.logger.debug({
+        at: "PriceClient#getPriceByAddress",
+        message: `Looking up prices via ${priceFeed.name}.`,
+      });
+      try {
+        prices = await priceFeed.getPricesByAddress(addresses, currency);
+        if (prices.length === 0) {
+          throw Error("Zero-length response received");
+        }
+      } catch (err) {
+        this.logger.debug({
+          at: "PriceClient#requestPrices",
+          message: `Price lookup against ${priceFeed.name} failed (${err}).`,
+          tokens: addresses,
+        });
+        continue; // Failover to the next price feed...
```

Comment:
nit: is this continue redundant? Won't it loop anyway?

Reply:
Yeah; and there's no way to break out of the loop when a successful retrieval occurs. That's pretty critical...